modules/smtp/modules.php

```diff
@@ -2291,3 +2452,102 @@ function recip_count_check($headers, $omod) {
         Hm_Msgs::add('ERRMessage contains more than the maximum number of recipients, proceed with caution');
     }
 }}
+
+/**
+ * @subpackage smtp/functions
+ */
+if (!hm_exists('send_scheduled_message')) {
+function send_scheduled_message($handler, $imap, $msg, $server_id) {
```

Comment:
Can you combine this with the send_scheduled_messages handler method, so we don't repeat so much code?

Reply:
For the concern about code repetition. the two methods have slightly different execution contexts. one handles message rescheduling and performs specific tasks related to IMAP connection and cache fetching, while the other is more geared towards sending messages at a specific time based on message headers. planning. I have tried several ways to merge but I think that merging these two methods could complicate the logic and could make the code more difficult to maintain.

Reply:
I fail to see how repeating 49 lines of code will be easier to maintain in the future. This is the code I see that is 100% repeated:
```
                        if (new DateTime($msg_headers['X-Schedule']) <= new DateTime()) {
                            $profile = Hm_Profiles::get($msg_headers['X-Profile-ID']);
                            if (!$profile) {
                                $profiles = Hm_Profiles::search('server', $imap_details['server']);

                                if (!$profiles) {
                                    Hm_Debug::add(sprintf('ERRCannot find profiles corresponding with IMAP server: %s', $imap_details['server']));
                                    return;
                                }
                                $profile = $profiles[0];
                            }

                            $smtp = Hm_SMTP_List::connect($profile['smtp_id'], false);

                            if (smtp_authed($smtp)) {
                                if (isset($msg_headers['X-Delivery'])) {
                                    $from_params = 'RET=HDRS';
                                    $recipients_params = 'NOTIFY=SUCCESS,FAILURE';
                                } else {
                                    $from_params = '';
                                    $recipients_params = '';
                                }

                                $recipients = [];
                                foreach (['To', 'Cc', 'Bcc'] as $fld) {
                                    if (array_key_exists($fld, $msg_headers)) {
                                        $recipients = array_merge($recipients, Hm_MIME_Msg::find_addresses($msg_headers[$fld]));
                                    }
                                }

                                $msg_content = $imap->get_message_content($msg['uid'], 0);
                                $from = process_address_fld($msg_headers['From']);

                                $err_msg = $smtp->send_message($from[0]['email'], $recipients, $msg_content, $from_params, $recipients_params);

                                if (!$err_msg) {
                                    if ($imap->message_action('DELETE', [$msg['uid']])) {
                                        $imap->message_action('EXPUNGE', [$msg['uid']]);
                                    }
                                    save_sent_msg($handler, $server_id, $imap, $imap_details, $msg_content, $msg['uid'], false);
                                    return true;
                                }
                            }
                        }
                    } catch (Exception $e) {
                        Hm_Debug::add(sprintf('ERRCannot send message: %s', $msg_headers['subject']));
                    }
                    return;
                }}
```
Please find a way to abstract it and reuse.

Reply:
Hello @kroky  I am making changes. please review

Reply:
Yes, it is better. You can move all the code in send_message_if_scheduled to send_scheduled_message and remove send_message_if_scheduled. There is no need to call 2 separate functions with exactly the same arguments and make one call the other without doing anything else - just use one function.

Reply:
@kroky I just made some changes. Please review

Reply:
Yes, it is fine.